[RFC] vim-patch:8.2.0629

Billy4195 opened the pull request:
#### vim-patch:8.2.0629: setting a boolean option to v:false does not work

Problem:    Setting a boolean option to v:false does not work.
Solution:   Do not use the string representation of the value. (Christian
            Brabandt, closes vim/vim#5974)
https://github.com/vim/vim/commit/65d032c779a43b767497e15e6a32d04a6a8fa65d

cc #12174

janlazo commented on src/nvim/eval.c:
```diff
@@ -1833,12 +1833,15 @@ static char_u *ex_let_one(char_u *arg, typval_T *const tv,
 
       varnumber_T n = tv_get_number(tv);
-      const char *s = tv_get_string_chk(tv);  // != NULL if number or string.
+      if (tv->v_type != VAR_SPECIAL) {
```
`VAR_BOOL` is relevant change but not ported here. Need vim-patch:8.2.0111.

Billy4195 replied:
OK, I'll add it.

Billy4195 replied:
@janlazo after surveyed the vim-patch:8.2.0111, I found it needs more patches (patches before 8.2.0111) to be ported to finish this PR.
Should I need to add all of them? or this PR could be done without adding those patches.

janlazo replied:
It will be harder to port 8.2.0111 and related patches if this PR is merged now.

Billy4195 replied:
I got it, I will take some time to trace back and add patches.

Billy4195 committed:
vim-patch:8.2.0111: VAR_SPECIAL is also used for booleans

Problem:    VAR_SPECIAL is also used for booleans.
Solution:   Add VAR_BOOL for better type checking.
https://github.com/vim/vim/commit/9b4a15d5dba354d2e1e02871470bad103f34769a

Billy4195 committed:
vim-patch:8.2.0629: setting a boolean option to v:false does not work

Problem:    Setting a boolean option to v:false does not work.
Solution:   Do not use the string representation of the value. (Christian
            Brabandt, closes vim/vim#5974)
https://github.com/vim/vim/commit/65d032c779a43b767497e15e6a32d04a6a8fa65d

teto commented:
when porting the patch 8.2.0111 patch, you missed the `tv_op` changes that now reside at https://github.com/neovim/neovim/blob/97bcab8f5eeb8c5b2627fdfc3adad2682bb3681e/src/nvim/eval/executor.c#L31 : we shoud not be able to run operations against booleans.

teto commented:
@Billy4195 I would like to merge this if you could just address the single comment :)

Billy4195 commented:
@teto Done. 
Sorry for late update, I'm working on my thesis these days.

> when porting the patch 8.2.0111 patch, you missed the `tv_op` changes that now reside at
> 
> https://github.com/neovim/neovim/blob/97bcab8f5eeb8c5b2627fdfc3adad2682bb3681e/src/nvim/eval/executor.c#L31
> 
> : we shoud not be able to run operations against booleans.

teto commented:
looks like we merged some patches that renamed some functions so it stopped building. Good luck with your thesis (you don't need it I am sure :) )

Billy4195 commented:
@teto I have fixed the problem. 
Should I add more test cases to cover these patches?
